### PR TITLE
[Merged by Bors] - feat: port Data.Nat.PSub

### PR DESCRIPTION
**Pull request by zeramorphic**
mathlib3 SHA: 62a5626868683c104774de8d85b9855234ac807c

I'm not sure if the file should be called `Psub.lean` or `PSub.lean` - mathport chooses the former.

### CODE REVIEW

**Comment by semorrison** [comment]
Please rename to `PSub`.

File: Mathlib/Data/Nat/Psub.lean [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Mathlib.Data.Option.Basic
+import Mathlib.Data.Nat.Basic
+
+/-!
+# Partial predecessor and partial subtraction on the natural numbers
+
+The usual definition of natural number subtraction (`nat.sub`) returns 0 as a "garbage value" for
```
Comment (semorrison) [inline Mathlib/Data/Nat/Psub.lean:12]:
```suggestion
The usual definition of natural number subtraction (`Nat.sub`) returns 0 as a "garbage value" for
```

File: Mathlib/Data/Nat/Psub.lean [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+`a - b` when `a < b`. Similarly, `nat.pred 0` is defined to be `0`. The functions in this file
```
Comment (semorrison) [inline Mathlib/Data/Nat/Psub.lean]:
```suggestion
`a - b` when `a < b`. Similarly, `Nat.pred 0` is defined to be `0`. The functions in this file
```

File: Mathlib/Data/Nat/Psub.lean [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+wrap the result in an `option` type instead:
+
+## Main definitions
+
+- `nat.ppred`: a partial predecessor operation
+- `nat.psub`: a partial subtraction operation
```
Comment (semorrison) [inline Mathlib/Data/Nat/Psub.lean]:
```suggestion
- `Nat.ppred`: a partial predecessor operation
- `Nat.psub`: a partial subtraction operation
```

File: Mathlib/Data/Nat/PSub.lean [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+# Partial predecessor and partial subtraction on the natural numbers
+
+The usual definition of natural number subtraction (`nat.sub`) returns 0 as a "garbage value" for
+`a - b` when `a < b`. Similarly, `nat.pred 0` is defined to be `0`. The functions in this file
+wrap the result in an `option` type instead:
+
+## Main definitions
+
+- `nat.ppred`: a partial predecessor operation
+- `nat.psub`: a partial subtraction operation
```
Comment (semorrison) [inline Mathlib/Data/Nat/PSub.lean:19]:
```suggestion
The usual definition of natural number subtraction (`Nat.sub`) returns 0 as a "garbage value" for
`a - b` when `a < b`. Similarly, `Nat.pred 0` is defined to be `0`. The functions in this file
wrap the result in an `Option` type instead:

## Main definitions

- `Nat.ppred`: a partial predecessor operation
- `Nat.psub`: a partial subtraction operation
```

File: Mathlib/Data/Nat/PSub.lean [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+-/
+
+
+namespace Nat
+
+/-- Partial predecessor operation. Returns `ppred n = some m`
+  if `n = m + 1`, otherwise `none`. -/
+@[simp]
```
Comment (semorrison) [inline Mathlib/Data/Nat/PSub.lean:28]: Please remove `@[simp]` here and on `psub`. You may need to write the simp lemmas for the branches yourself.

Putting `@[simp]` on a definition by matching has changed behaviour in Lean 4.

File: Mathlib/Data/Nat/PSub.lean [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+-- Porting note: `contradiction` required an `intro` for the goals
+-- `ppred (n + 1) = none → n + 1 = 0` and `n + 1 = 0 → ppred (n + 1) = none`
+
+@[simp]
+theorem ppred_eq_none : ∀ {n : ℕ}, ppred n = none ↔ n = 0
+  | 0 => by simp
+  | n + 1 => by constructor <;> intro <;> contradiction
+#align nat.ppred_eq_none Nat.ppred_eq_none
+
+-- Porting note: `simp` eagerly expanded `psub` and `ppred` so it couldn't apply the `simp` lemmas
```
Comment (semorrison) [inline Mathlib/Data/Nat/PSub.lean:67]: See above.

File: Mathlib/Data/Nat/PSub.lean [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+theorem psub_eq_sub {m n} (h : n ≤ m) : psub m n = some (m - n) :=
+  psub_eq_some.2 <| Nat.sub_add_cancel h
+#align nat.psub_eq_sub Nat.psub_eq_sub
+
+-- Porting note: we only have the simp lemma `Option.bind_some` which uses `Option.bind` not `>>=`
+
+theorem psub_add (m n k) :
+    psub m (n + k) = (do
+      let x ← psub m n
+      psub x k) :=
```
Comment (semorrison) [inline Mathlib/Data/Nat/PSub.lean:101]:
```suggestion
    psub m (n + k) = (do psub (← psub m n) k) :=
```

File: Mathlib/Data/Nat/PSub.lean [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+/-- Partial subtraction operation. Returns `psub m n = some k`
+  if `m = n + k`, otherwise `none`. -/
+@[simp]
+def psub (m : ℕ) : ℕ → Option ℕ
+  | 0 => some m
+  | n + 1 => psub m n >>= ppred
+#align nat.psub Nat.psub
+
+-- Porting note: mathport failed to align `option.get_or_else` with `Option.getD`
+
```
Comment (semorrison) [inline Mathlib/Data/Nat/PSub.lean:43]:
```suggestion
```

**Comment by zeramorphic** [comment]
Thanks for your comments, I think I've made all of the relevant changes.

**Comment by semorrison** [comment]
Great, thanks. It's helpful if in mathlib PR reviews, if the author can click "resolve conversation" on everything they've dealt with. Because we have multiple reviewers dealing with things, it's great if someone coming in can quickly see what is and isn't done.

bors merge

**Comment by zeramorphic** [comment]
That makes sense, thanks - I'm not very used to collaborating on projects like this just yet.

**Pull request closed without merging**